hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/opt/prunability/RelPrunabilityTest.java

```diff
@@ -0,0 +1,216 @@
+import com.hazelcast.jet.sql.impl.opt.OptimizerTestSupport;
+import com.hazelcast.jet.sql.impl.opt.metadata.HazelcastRelMetadataQuery;
+import com.hazelcast.jet.sql.impl.opt.physical.AggregateAccumulateByKeyPhysicalRel;
+import com.hazelcast.jet.sql.impl.opt.physical.AggregateCombineByKeyPhysicalRel;
+import com.hazelcast.jet.sql.impl.opt.physical.CalcPhysicalRel;
+import com.hazelcast.jet.sql.impl.opt.physical.FullScanPhysicalRel;
+import com.hazelcast.jet.sql.impl.opt.physical.IndexScanMapPhysicalRel;
+import com.hazelcast.jet.sql.impl.opt.physical.JoinNestedLoopPhysicalRel;
+import com.hazelcast.jet.sql.impl.opt.physical.PhysicalRel;
+import com.hazelcast.jet.sql.impl.opt.physical.UnionPhysicalRel;
+import com.hazelcast.jet.sql.impl.schema.HazelcastTable;
+import com.hazelcast.jet.sql.impl.validate.types.HazelcastTypeFactory;
+import com.hazelcast.jet.sql.impl.validate.types.HazelcastTypeUtils;
+import com.hazelcast.map.IMap;
+import com.hazelcast.sql.impl.extract.QueryPath;
+import com.hazelcast.sql.impl.schema.TableField;
+import com.hazelcast.sql.impl.schema.map.MapTableField;
+import com.hazelcast.sql.impl.type.QueryDataType;
+import org.apache.calcite.rel.metadata.RelMetadataQuery;
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.rex.RexLiteral;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.sql.type.SqlTypeName;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+import java.util.List;
+import java.util.Map;
+
+import static com.hazelcast.sql.impl.extract.QueryPath.KEY;
+import static com.hazelcast.sql.impl.extract.QueryPath.VALUE;
+import static com.hazelcast.sql.impl.schema.map.MapTableUtils.getPartitionedMapIndexes;
+import static com.hazelcast.sql.impl.type.QueryDataType.BIGINT;
+import static com.hazelcast.sql.impl.type.QueryDataType.VARCHAR;
+import static java.util.Arrays.asList;
+import static java.util.Collections.emptyMap;
+import static java.util.Collections.singletonList;
+import static org.junit.Assert.assertEquals;
+
+public class RelPrunabilityTest extends OptimizerTestSupport {
+    private static final RelDataType REL_TYPE_BIGINT = HazelcastTypeUtils.createType(
+            HazelcastTypeFactory.INSTANCE,
+            SqlTypeName.BIGINT,
+            true);
+
+    private HazelcastTable table;
+    private HazelcastRelMetadataQuery query;
+
+    @BeforeClass
+    public static void beforeClass() throws Exception {
+        initialize(1, null);
+    }
+
+    @Before
+    public void before() throws Exception {
+        table = partitionedTable(
+                "m",
+                asList(
+                        mapField(KEY, BIGINT, QueryPath.KEY_PATH),
+                        mapField(VALUE, VARCHAR, QueryPath.VALUE_PATH)),
+                10);
+    }
+
+    @Test
+    public void test_fullScan() {
+        PhysicalRel root = optimizePhysical(
+                "SELECT __key FROM m WHERE __key = 10 AND this IS NOT NULL",
```

Comment:
partition pruning for `__key` is less important than for `AttributePartitioningStrategy`. Do we have similar tests for `AttributePartitioningStrategy`? They can have many cases, eg: 1 partitioning attribute vs many attributes; not all attributes of key used vs all attributes used.

Reply:
Legit request, will complement test

Reply:
Done - https://github.com/hazelcast/hazelcast/pull/24813/commits/7f2eac52c581253ea78f2b469656dc1e3f668828

Reply:
tests with `__key` were fine, just less important, so we should keep them as regression tests for that case.

I also do not see explicit test with residual filters in `RelPrunabilityTest` (test_union has additional filters but it is not the main purpose of that test).